Formatting of relative and absolute dates using Intl.* APIs

niik opened the pull request:
<!--
What GitHub Desktop issue does this PR address? (for example, #1234)
-->

Closes #13943 

## Description
<!--
A summary of the changes made along with any other information that would be helpful to a reviewer such as potential tradeoffs or alternative approaches you considered.
-->
This adds ~~region-aware~~ formatting for relative and absolute dates using the [`Intl`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl) namespace. This builds on #14149, #14123 and [app.getLocaleCountryCode](https://www.electronjs.org/docs/latest/api/app#appgetlocalecountrycode)

Note that this is not about localizing strings, the interface will remain US-english, ~~it will however apply the correct (hopefully) regional settings for formatting numbers. Most notably right now this means that users like myself who have their OS configured to display 24-hour format will now.~~

**Note that this is feature-flagged and currently only enabled for dev**.

cc @tsvetilian-ty as author of #13965 

### Screenshots

<!--
If this PR touches the UI layer of the app, please include screenshots or animated gifs to show the changes.
-->

<img width="329" alt="image" src="https://user-images.githubusercontent.com/634063/158350973-04f3495b-2767-4248-ab58-b26e7e9e1fdd.png">

For reference these are my regional settings on Windows and macOS (they resolve to `en-SE` on both platforms)

<img width="718" alt="image" src="https://user-images.githubusercontent.com/634063/158351255-7b486c6b-595d-44ca-b244-03ca95636b95.png">

<img width="780" alt="image" src="https://user-images.githubusercontent.com/634063/158351303-50aa9425-287c-4a7d-9dd9-e0b6bef5c3e1.png">


## Release notes

<!--
You can leave this blank if you're not sure.
If you don't believe this PR needs to be mentioned in the release notes, write "Notes: no-notes".
-->

Notes:

niik committed:
IPC method to grab the AppleLocale setting

niik committed:
Parse and store user locale

niik committed:
Use user locale when formatting absolute dates

niik committed:
Bump win32-user-locale

niik committed:
feature flag locale aware date formatting

niik committed:
Typo

niik committed:
Add loader for resolving 'bindings' path at compile time

niik committed:
Cleanup logging

niik committed:
Be less specific about locale relating to dates

niik committed:
Be a good citizen and new that class up

niik committed:
All kinds of formatting

niik committed:
Log when unable to init formatter

niik committed:
Merge branch 'development' into locale-aware-formatting

niik committed:
Region-aware relative time format

niik committed:
:book:

niik committed:
Add a region-aware number formatter

niik committed:
Replace a bunch of similar patterns with a format function

niik committed:
Generic formatCount method

niik committed:
More stragglers converted to formatCount

niik committed:
Everything is terrible

niik committed:
Merge branch 'locale-aware-formatting' into region-aware-number-formatting

sergiou87 commented on app/jest.unit.config.js:
```diff
@@ -23,5 +23,7 @@ module.exports = {
   coverageReporters: ['text-summary', 'json', 'html', 'cobertura'],
   // For now, @github Node modules required to be transformed by jest-esm-transformer
-  transformIgnorePatterns: ['node_modules/(?!(@github))'],
+  transformIgnorePatterns: [
+    'node_modules/(?!(@github|bcp-47|is-decimal|is-alphabetical|is-alphanumerical))',
+  ],
```
> [Everything is terrible](https://github.com/desktop/desktop/pull/14173/commits/daa48753397abb86114d65b4b62630f45e23e8be)

:its-true-all-of-it:

niik committed:
Add a little helpful tooltip to the compact ahead/behind number

niik committed:
Don't have to accept a unit in formatCommitCount, it's supposed to be a shorthand

niik committed:
Merge branch 'development' into locale-aware-formatting

niik committed:
Merge branch 'locale-aware-formatting' into region-aware-number-formatting

niik commented:
@sergiou87 Merge conflict, need a little ✅ again please

niik committed:
Use getLocaleCountryCode instead of custom logic

niik committed:
Case insensitive

niik committed:
:art: cleanup

niik committed:
Merge branch 'locale-aware-formatting' into region-aware-number-formatting

niik committed:
Whoops

niik committed:
Merge branch 'locale-aware-formatting' into region-aware-number-formatting

niik committed:
Revert to old syntax

niik committed:
Revert to old syntax

niik committed:
Don't need this any more

niik committed:
Don't need this any more

niik committed:
Shared caching of all Intl formatters

niik committed:
:art: cleanup

sergiou87 commented:
Sorry @niik I didn't review it yet, only left that dumb comment 😅 I'll review it tomorrow first thing in the morning!

niik commented:
> Sorry @niik I didn't review it yet, only left that dumb comment 😅 I'll review it tomorrow first thing in the morning!

No rush!

niik committed:
Format percent numbers

niik committed:
Convert formatBytes to use formatNumber and rename file to match format* functions

niik committed:
Fix imports

sergiou87 commented on app/src/lib/formatting-locale.ts:
```diff
@@ -0,0 +1,30 @@
+let formattingLocale: string | undefined = undefined
+
+/**
+ * Get the current user localle or undefined if not set or invalid
```
```suggestion
 * Get the current user locale or undefined if not set or invalid
```

niik committed:
Bring in existing tests and update them with the new reality

sergiou87 reviewed (commented):
As discussed in Slack, it's not picking the 24h format for me, but seems to be an issue with `Intl` rather than with your implementation:

![image](https://user-images.githubusercontent.com/1083228/158575531-28f099c2-579d-4c67-9aa8-438d025f7f90.png)

![image](https://user-images.githubusercontent.com/1083228/158575580-5dc37f75-de89-4371-b574-632d766491b4.png)

niik committed:
Update app/src/lib/formatting-locale.ts

Co-authored-by: Sergio Padrino <sergio.padrino@gmail.com>

niik committed:
Bring this back to dev for now

niik committed:
Use en-US-POSIX instead of en-US

en-US-POSIX gives us something very close to the standard behavior of number.toString() et al, i.e. an invariant culture

niik committed:
Typo

niik commented:
Alright, after significant investigation here we've determined that we're not quite ready to jump on the regional-aware train just yet. There are some notable en-* locales that isn't defined in the Unicode Common Locale Data Repository such as en-ES, en-NO, en-FR, etc.

We have ideas for how to address this though and in the meantime we'll be using the `en-US-POSIX` locale by default. The POSIX variant of `en-US` is essentially US-english but without number grouping (i.e it's essentially `.toString()`).

Given the recent discoveries I'm moving the feature flag back to `dev` and intend to tackle the locale resolution in a separate PR.

@sergiou87 Another look when you've got time please?

niik committed:
Merge branch 'locale-aware-formatting' into region-aware-number-formatting

niik committed:
Format title using formatNumber

Co-Authored-By: tidy-dev <75402236+tidy-dev@users.noreply.github.com>

niik committed:
:art: typo

Co-Authored-By: tidy-dev <75402236+tidy-dev@users.noreply.github.com>

niik committed:
Merge branch 'region-aware-number-formatting' into shared-formatter-caching

niik committed:
We're using en-US-POSIX now, no grouping

niik committed:
Merge branch 'region-aware-number-formatting' into shared-formatter-caching

sergiou87 reviewed (dismissed):
Looks and works good! :shipit:

niik committed:
Merge branch 'development' into region-aware-number-formatting

niik committed:
Merge branch 'development' into locale-aware-formatting

angusdev committed:
Issue 15284 Toggle Diff Display Mode and Hide Whitespace Changes

niik committed:
Merge pull request #14176 from desktop/shared-formatter-caching

Shared formatter caching

niik committed:
Merge branch 'locale-aware-formatting' into region-aware-number-formatting

niik committed:
Merge branch 'development' into region-aware-number-formatting

niik committed:
Merge fail

niik committed:
Merge pull request #14174 from desktop/region-aware-number-formatting

Number formatting using Intl.* APIs

niik committed:
Revert "Issue 15284 Toggle Diff Display Mode and Hide Whitespace Changes"

This reverts commit 324986874ddfd9e1c4ba57c0b04fb8df7e96adfe.

niik committed:
Lint

johnnyjeannatasha commented on app/src/lib/format-relative.ts:
```diff
@@ -1,21 +1,30 @@
 import mem from 'mem'
```
Importing. Mem

johnnyjeannatasha commented on app/src/lib/format-relative.ts:
```diff
@@ -1,21 +1,30 @@
 import mem from 'mem'
 import QuickLRU from 'quick-lru'
+import { getFormattingLocales } from './formatting-locale'
```
Importing loca

johnnyjeannatasha commented on app/src/lib/format-relative.ts:
```diff
@@ -1,21 +1,30 @@
 import mem from 'mem'
 import QuickLRU from 'quick-lru'
```
Importing lru

johnnyjeannatasha commented on app/src/lib/format-number.ts:
```diff
@@ -0,0 +1,31 @@
+ *
+ * See Intl.NumberFormat for more information
+ */
+export const formatNumber = (num: number, options?: Intl.NumberFormatOptions) =>
```
Exporting contact formatting

johnnyjeannatasha commented on app/src/ui/branches/push-branch-commits.tsx:
```diff
@@ -95,28 +79,29 @@ export class PushBranchCommits extends React.Component<
-            and open a pull request?
+            Would you like to publish <Ref>{branch.name}</Ref> now and open a
+            pull request?
           </p>
```
Pull request

johnnyjeannatasha commented on app/src/ui/history/merge-call-to-action-with-conflicts.tsx:
```diff
@@ -14,6 +14,7 @@ import {
 import { getMergeOptions, updateRebasePreview } from '../lib/update-branch'
 import { MultiCommitOperationKind } from '../../models/multi-commit-operation'
 import { RebasePreview } from '../../models/rebase'
```
Import

johnnyjeannatasha replied:
Rebase

johnnyjeannatasha commented on app/src/ui/history/merge-call-to-action-with-conflicts.tsx:
```diff
@@ -252,14 +253,12 @@ export class MergeCallToActionWithConflicts extends React.Component<
-    const pluralized = this.commitCount === 1 ? 'commit' : 'commits'
-
     if (this.state.selectedOperation === MultiCommitOperationKind.Rebase) {
       return (
```
Return(

johnnyjeannatasha commented on app/src/ui/history/merge-call-to-action-with-conflicts.tsx:
```diff
@@ -14,6 +14,7 @@ import {
 import { getMergeOptions, updateRebasePreview } from '../lib/update-branch'
 import { MultiCommitOperationKind } from '../../models/multi-commit-operation'
 import { RebasePreview } from '../../models/rebase'
+import { formatCommitCount } from '../../lib/format-commit-count'
```
Import from format commit

johnnyjeannatasha commented on app/src/ui/history/merge-call-to-action-with-conflicts.tsx:
```diff
@@ -14,6 +14,7 @@ import {
 import { getMergeOptions, updateRebasePreview } from '../lib/update-branch'
 import { MultiCommitOperationKind } from '../../models/multi-commit-operation'
```
Import from base

johnnyjeannatasha commented on app/src/ui/history/merge-call-to-action-with-conflicts.tsx:
```diff
@@ -252,14 +253,12 @@ export class MergeCallToActionWithConflicts extends React.Component<
       return null
     }
 
-    const pluralized = this.commitCount === 1 ? 'commit' : 'commits'
```
Constant plurized

johnnyjeannatasha commented on app/src/lib/format-commit-count.ts:
```diff
@@ -0,0 +1,15 @@
+import { formatNumber } from './format-number'
```
Import from format number

tidy-dev commented:
superseded by https://github.com/desktop/desktop/pull/21836
